stubs/SQLAlchemy/sqlalchemy/engine/base.pyi

```diff
@@ -1,61 +1,97 @@
-    def invalidated(self): ...
+    def invalidated(self) -> bool: ...
     @property
-    def connection(self): ...
+    def connection(self) -> DBAPIConnection: ...
     def get_isolation_level(self): ...
     @property
     def default_isolation_level(self): ...
     @property
     def info(self): ...
     def connect(self, close_with_result: bool = ...): ...  # type: ignore[override]
-    def invalidate(self, exception: Any | None = ...): ...
+    def invalidate(self, exception: Exception | None = ...) -> None: ...
     def detach(self) -> None: ...
-    def begin(self): ...
-    def begin_nested(self): ...
-    def begin_twophase(self, xid: Any | None = ...): ...
+    def begin(self) -> Transaction | None: ...
+    def begin_nested(self) -> Transaction | None: ...
+    def begin_twophase(self, xid: Any | None = ...) -> TwoPhaseTransaction: ...
     def recover_twophase(self): ...
     def rollback_prepared(self, xid, recover: bool = ...) -> None: ...
     def commit_prepared(self, xid, recover: bool = ...) -> None: ...
-    def in_transaction(self): ...
-    def in_nested_transaction(self): ...
-    def get_transaction(self): ...
-    def get_nested_transaction(self): ...
+    def in_transaction(self) -> bool: ...
+    def in_nested_transaction(self) -> bool: ...
+    def get_transaction(self) -> Transaction | None: ...
+    def get_nested_transaction(self) -> Transaction | None: ...
     def close(self) -> None: ...
-    def scalar(self, object_, *multiparams, **params): ...
+    @overload
+    def scalar(
+        self,
+        object_: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
+        *multiparams: Mapping[str, Any],
+        **params: Any,
+    ) -> Any: ...
+    @overload
+    def scalar(self, object_: str, *multiparams: Any | tuple[Any, ...] | Mapping[str, Any], **params: Any) -> Any: ...
     def scalars(self, object_, *multiparams, **params): ...
-    def execute(self, statement, *multiparams, **params): ...
-    def exec_driver_sql(self, statement, parameters: Any | None = ..., execution_options: Any | None = ...): ...
-    def transaction(self, callable_, *args, **kwargs): ...
-    def run_callable(self, callable_, *args, **kwargs): ...
+    @overload  # type: ignore[override]
+    def execute(
+        self,
+        statement: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
+        *multiparams: Mapping[str, Any],
+        **params,
+    ) -> CursorResult: ...
+    @overload
+    def execute(self, statement: str, *multiparams: Any | tuple[Any, ...] | Mapping[str, Any], **params) -> CursorResult: ...
+    def exec_driver_sql(self, statement: str, parameters: Any | None = ..., execution_options: Any | None = ...): ...
+    # TODO:
+    # def transaction(self, callable_: Callable[Concatenate[Connection, _P], _T], *args: _P.args, **kwargs: _P.kwargs) -> _T | None: ...
+    def transaction(self, callable_: Callable[..., _T], *args: Any, **kwargs: Any) -> _T | None: ...
```

Comment:
How can this return `None`, if `run_callable()` can't? The `except:` in he implementation calls `utils.safe_reraise()`, which I would expect to raise an exception.